File: ts/128/103.ts
import { TreeNode } from '../class'

// BFS
function zigzagLevelOrder(root: TreeNode | null): number[][] {
  let dir: boolean = true
  const ans: number[][] = []
  if (!root) return ans
  const queue: TreeNode[] = [root]
  while (queue.length) {
    const item: number[] = []
    const len = queue.length
    for (let i = 0; i < len; i++) {
      const node = queue.shift()
      if (node) {
        if (dir) {
          item.push(node.val)
        } else {
          item.unshift(node.val)
        }
        if (node.left) queue.push(node.left)
        if (node.right) queue.push(node.right)
      }
    }
    ans.push(item)
    dir = !dir
  }

  return ans
};

File: py3/128/103.py
from collections import deque
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        ans, que = [], deque()
        if root:
            flag = True
            que.append(root)
            while que:
                items, len_que = [], len(que)
                for i in range(len_que):
                    node = que.popleft()
                    items.append(node.val)

                    if node.left:
                        que.append(node.left)
                    if node.right:
                        que.append(node.right)
                if flag:
                    ans.append(items)
                else:
                    ans.append(items[::-1])

                flag = not flag

        return ans